kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/partition/FixedLongRangeFieldPartitioner.java

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright 2013 Cloudera Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kitesdk.data.spi.partition;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import com.google.common.collect.DiscreteDomains;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+import org.kitesdk.data.spi.FieldPartitioner;
+import org.kitesdk.data.spi.predicates.Exists;
+import org.kitesdk.data.spi.predicates.In;
+import org.kitesdk.data.spi.predicates.Predicates;
+import org.kitesdk.data.spi.predicates.Range;
+import org.kitesdk.data.spi.predicates.Ranges;
+
+@edu.umd.cs.findbugs.annotations.SuppressWarnings(value={
+    "NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE",
+    "SE_COMPARATOR_SHOULD_BE_SERIALIZABLE"},
+    justification="False positive due to generics.")
+@Immutable
+public class FixedLongRangeFieldPartitioner extends FieldPartitioner<Long, Long> {
+
+  private final long size;
+
+  public FixedLongRangeFieldPartitioner(String sourceName, long size) {
+    this(sourceName, null, size);
+  }
+
+  public FixedLongRangeFieldPartitioner(String sourceName, @Nullable String name,
+      long size) {
+    super(sourceName, (name == null ? sourceName + "_range" : name),
+        Long.class, Long.class);
+    this.size = size;
+    Preconditions.checkArgument(size > 0,
+        "Size of range buckets is not positive: %s", size);
+  }
+
+  @Override
+  public Long apply(Long value) {
+    return Math.round(Math.floor(value / ((double) size))) * size;
+  }
+
+  @Override
+  public Predicate<Long> project(Predicate<Long> predicate) {
+    if (predicate instanceof Exists) {
+      return Predicates.exists();
+    } else if (predicate instanceof In) {
+      return ((In<Long>) predicate).transform(this);
+    } else if (predicate instanceof Range) {
+      // must use a closed range:
+      //   if this( 5 ) => 10 then this( 6 ) => 10, so 10 must be included
+      return Ranges.transformClosed(
+          Ranges.adjustClosed((Range<Long>) predicate,
+              DiscreteDomains.longs()), this);
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public Predicate<Long> projectStrict(Predicate<Long> predicate) {
+    if (predicate instanceof Exists) {
+      return Predicates.exists();
+    } else if (predicate instanceof In) {
+      return ((In<Long>) predicate).transform(this);
```

Comment:
I think this strict projection is incorrect. The only way a value should be matched here is if all of the inputs to this partitioner that produce this value are matched by the predicate. If the partitioner has size 5 and its strict predicate returns 10, then the original predicate must have matched 10, 11, 12, 13, and 14: all of the values that get mapped to 10. Because these are integers, we can actually check them, though returning null is fine with me because it is a pain to check all possible values. Returning null will make Kite apply the original predicate when reading the data.

On the other hand, ranges are easier to reason about and should be included. You can check each endpoint individually: if `(start % size) == 0` then the projected range should start there, otherwise start at the next bucket. If `((end + 1) % size) == 0` then include the bucket for end, otherwise the one before that. Then construct a closed range from the bucket endpoints and be sure to check that start < end. Hopefully that makes sense!

This method will need a test as well.